### PR TITLE
Add gitlabMergeCommitSha to Fix #1046

### LINKED ISSUE #1046: Jenkins Gitlab squash commit error stderr: fatal: ambiguous argument '^{commit}': unknown revision or path not in the working tree.  Use '--' to separate paths from revisions, like this: 'git <command> [<revision>...] -- [<file>...]'
When we merge with squash commit Jenkins fail with the following error
` 
```09:27:56  hudson.plugins.git.GitException: Command "git rev-parse d8fd53f8c5c1522678cfae1af81f2c61eb1ec8a3^{commit}" returned status code 128:
09:27:56  stdout: d8fd53f8c5c1522678cfae1af81f2c61eb1ec8a3^{commit}
09:27:56  
09:27:56  stderr: fatal: ambiguous argument 'd8fd53f8c5c1522678cfae1af81f2c61eb1ec8a3^{commit}': unknown revision or path not in the working tree.
09:27:56  Use '--' to separate paths from revisions, like this:
09:27:56  'git <command> [<revision>...] -- [<file>...]```

However when we merge without squash commit, pipeline runs successful 
This issue happen when we checkout using gitlab plugin and below is the Jenkins checkout script

  checkout changelog: true, poll: true, scm: [
            $class: 'GitSCM',
            branches: [[name: "origin/${env.gitlabSourceBranch}"]],
            doGenerateSubmoduleConfigurations: false,
            extensions: [],
            submoduleCfg: [],
            userRemoteConfigs: [[
            credentialsId: "gitlab-ssh",
            url: gitlabTargetRepoSshUrl
            ]]
        ]
```
```
Please can you advise?

### PR DESCRIPTION
**Pull request by MrSelfDestruction**
Hello, 

This PR fix the following problems:
- https://github.com/jenkinsci/gitlab-plugin/issues/1046
- https://issues.jenkins.io/browse/JENKINS-61964

If parameter `gitlabMergeCommitSha` is found in webhook, then Jenkins will check out git repository to squash/merge commit, not to the last commit from source branch.

Tested on the following environment:
- Jenkins 2.388 and 2.397
- GitLab Community Edition [15.11.2](https://gitlab.com/gitlab-org/gitlab-foss/-/tags/v15.11.2)

Closes #1046.

### CODE REVIEW

File: src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java]: Could you please use a non-empty dummy string as the test merge commit SHA in the test? Otherwise how do we know if it works?

**Review by krisstern — CHANGES_REQUESTED** [review]
Thanks @MrSelfDestruction for the code contribution!
Only one issue with this PR to follow up on.

File: src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java
Comment (MrSelfDestruction) [reply src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java]: > Could you please use a non-empty dummy string as the test merge commit SHA in the test? Otherwise how do we know if it works?

Hello, I have added this. Could you check?

File: src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java:134]:
```suggestion
        MergeRequest mergeRequest =
```

File: src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java
Comment (krisstern) [inline src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java:135]:
```suggestion
                new MergeRequest(MERGE_REQUEST_ID, MERGE_REQUEST_IID, MERGE_COMMIT_SHA, "", "", "", PROJECT_ID, PROJECT_ID, "", "");

```

**Review by krisstern — CHANGES_REQUESTED** [review]

File: src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java
Comment (MrSelfDestruction) [reply src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java:135]: This update was added by some linter. The CLI told me running `mvn spotless:apply` and I obeyed.

Comment (krisstern) [reply src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java:135]: No worries, I am still getting used to these `spotless` changes.

Comment (MrSelfDestruction) [reply src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java:135]: Is there anything else I should do?

Comment (krisstern) [reply src/test/java/com/dabsquared/gitlabjenkins/publisher/TestUtility.java:135]: None that I know of. I will pull the code locally and inspect it on my IDE later today before I merge it. But so far the code looks okay. I will probably be able to merge this PR if everything checks out and cut a new release later tonight.

**Comment by krisstern** [comment]
@MrSelfDestruction Just a reminder to not use the `master` branch for a feature pull request next time. It could potentially create all sorts of problems. Thanks!

**Review by krisstern — APPROVED** [review]

**Comment by toannh0104** [comment]
Hello @MrSelfDestruction, could you guide how to use/setup gitlabMergeCommitSha param please?. I'm going to upgrade gitlab-plugin version  
https://stackoverflow.com/questions/76771060/config-gitlab-webhook-with-gitlabmergecommitsha-parameter

**Comment by MrSelfDestruction** [comment]
> Hello @MrSelfDestruction, could you guide how to use/setup gitlabMergeCommitSha param please?. I'm going to upgrade gitlab-plugin version
> https://stackoverflow.com/questions/76771060/config-gitlab-webhook-with-gitlabmergecommitsha-parameter

Hello, 

gitlabMergeCommitSha used as fallback. See the following condition:

https://github.com/jenkinsci/gitlab-plugin/pull/1467/files#diff-66987b3e8de60fde15ec7ecaa12493bc5a00d2c9baa5b710a457a4d5f046fda8R247

If my memory serves me right, Gitlab sending gitlabMergeCommitSha when you are using squash for MR.

**Comment by toannh0104** [comment]
OK, thank you @MrSelfDestruction 
It seems that just need to upgrade to the newer version. I'm still overserve how that parameter work on merge with squash event

**Pull request merged**
